src/main/java/jcifs/smb1/netbios/NbtAddress.java

```diff
@@ -0,0 +1,880 @@
+            return;
+        }
+        long expiration = -1;
+        if( CACHE_POLICY != FOREVER ) {
+            expiration = System.currentTimeMillis() + CACHE_POLICY * 1000;
+        }
+        synchronized( ADDRESS_CACHE ) {
+            for( int i = 0; i < addrs.length; i++ ) {
+                CacheEntry entry = (CacheEntry)ADDRESS_CACHE.get( addrs[i].hostName );
+                if( entry == null ) {
+                    entry = new CacheEntry( addrs[i].hostName, addrs[i], expiration );
+                    ADDRESS_CACHE.put( addrs[i].hostName, entry );
+                } else {
+                    entry.address = addrs[i];
+                    entry.expiration = expiration;
+                }
+            }
+        }
+    }
+    static NbtAddress getCachedAddress( Name hostName ) {
+        if( CACHE_POLICY == 0 ) {
+            return null;
+        }
+        synchronized( ADDRESS_CACHE ) {
+            CacheEntry entry = (CacheEntry)ADDRESS_CACHE.get( hostName );
+            if( entry != null && entry.expiration < System.currentTimeMillis() &&
+                                                entry.expiration >= 0 ) {
+                entry = null;
+            }
+            return entry != null ? entry.address : null;
+        }
+    }
+
+    static NbtAddress doNameQuery( Name name, InetAddress svr )
+                                                    throws UnknownHostException {
+        NbtAddress addr;
+
+        if( name.hexCode == 0x1d && svr == null ) {
+            svr = CLIENT.baddr; // bit of a hack but saves a lookup
+        }
+        name.srcHashCode = svr != null ? svr.hashCode() : 0;
+        addr = getCachedAddress( name );
+
+        if( addr == null ) {
+            /* This is almost exactly like InetAddress.java. See the
+             * comments there for a description of how the LOOKUP_TABLE prevents
+             * redundant queries from going out on the wire.
+             */
+            if(( addr = (NbtAddress)checkLookupTable( name )) == null ) {
+                try {
+                    addr = CLIENT.getByName( name, svr );
+                } catch( UnknownHostException uhe ) {
+                    addr = UNKNOWN_ADDRESS;
+                } finally {
+                    cacheAddress( name, addr );
+                    updateLookupTable( name );
+                }
+            }
+        }
+        if( addr == UNKNOWN_ADDRESS ) {
+            throw new UnknownHostException( name.toString() );
+        }
+        return addr;
+    }
+
+    private static Object checkLookupTable( Name name ) {
+        Object obj;
+
+        synchronized( LOOKUP_TABLE ) {
+            if( LOOKUP_TABLE.containsKey( name ) == false ) {
+                LOOKUP_TABLE.put( name, name );
+                return null;
+            }
+            while( LOOKUP_TABLE.containsKey( name )) {
+                try {
+                    LOOKUP_TABLE.wait();
+                } catch( InterruptedException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)